File: dart/lib/src/packets/src/client/ps.dart
part of '../../packets.dart';

class PsPacket extends ClientPacket {
  /// [timestamp] is the time of the packet.
  /// This is identified in the packet as `UNIX`
  final DateTime timestamp;

  /// [params] is the configuration parameters of the packet.
  /// This is identified in the packet as `EXTRA+ARGS`
  final Map<String, dynamic> params;

  /// [PsPacket] is the configuration packet.
  ///
  /// This packet is part of the packet sent from the device to the server.
  ///
  /// Also, this packet only will be sent when `get_config` or `set_config` command is received.
  PsPacket({
    required this.timestamp,
    required this.params,
  });

  /// [fromPacket] creates a [PsPacket] from a string packet in the format of `Layrz Protocol v3`.
  static PsPacket fromPacket(String raw) {
    if (!raw.startsWith('<Ps>') || !raw.endsWith('</Ps>')) {
      throw ParseException('Invalid identification packet, should be <Ps>...</Ps>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    if (parts.length != 3) {
      throw MalformedException('Invalid packet parts, should have 3 parts');
    }

    int? receivedCrc = int.tryParse(parts[2], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 2).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    DateTime timestamp;
    try {
      timestamp = DateTime.fromMillisecondsSinceEpoch(int.parse(parts[0]) * 1000);
    } catch (e) {
      throw MalformedException('Invalid timestamp');
    }

    return PsPacket(
      timestamp: timestamp,
      params: Packet.parseExtraArgs(parts[1]),
    );
  }

  /// [toPacket] returns the packet in the format of `Layrz Protocol v3`.
  @override
  String toPacket() {
    String payload = '${(timestamp.millisecondsSinceEpoch / 1000).round()};';
    List<String> extraList = [];
    params.forEach((key, value) {
      if (value is bool) {
        extraList.add('$key:${value ? 'true' : 'false'}');
      } else if (value is int) {
        extraList.add('$key:$value');
      } else if (value is double) {
        extraList.add('$key:$value');
      } else {
        extraList.add('$key:$value');
      }
    });

    payload += "${extraList.join(',')};";

    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();

    return '<Ps>$payload$crc</Ps>';
  }

  @override
  String toString() => toPacket();

  @override
  PsPacket copyWith({
    DateTime? timestamp,
    Map<String, dynamic>? params,
  }) {
    return PsPacket(
      timestamp: timestamp ?? this.timestamp,
      params: params ?? this.params,
    );
  }
}

File: golang/v3/ps.go
package layrzprotocol

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Based on the `Layrz Protocol v2` specification. PsPacket is the Settings packet
// sent from the device to the server
type PsPacket struct {
	// Is the timestamp of the response packet
	Timestamp time.Time

	// Is the current configuration of the device
	Params *map[string]interface{}
}

// FromPacket is a method that converts a raw packet to a PaPacket
// based on the `Layrz Protocol v2` specification
//
// Returns an error if the packet is invalid
func (p *PsPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Ps>") || !strings.HasSuffix(*raw, "</Ps>") {
		return errors.New("invalid package, should be <Ps>...</Ps>")
	}

	*raw = strings.TrimPrefix(*raw, "<Ps>")
	*raw = strings.TrimSuffix(*raw, "</Ps>")

	rawCrc := (*raw)[len(*raw)-4:]
	*raw = (*raw)[:len(*raw)-4]

	receivedCrc, err := strconv.ParseUint(rawCrc, 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := calculateCrc([]byte(*raw))

	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	*raw = strings.TrimSuffix(*raw, ";")
	parts := strings.Split(*raw, ";")
	if len(parts) != 2 {
		return errors.New("invalid package, should contain 2 parts")
	}

	rawTimestamp, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return errors.New("cannot convert timestamp to integer")
	}

	p.Timestamp = time.Unix(rawTimestamp, 0)
	p.Params = parseArgs(parts[1])

	return nil
}

// ToPacket is a method that converts a PaPacket to a raw packet
// based on the `Layrz Protocol v2` specification
func (p *PsPacket) ToPacket() *string {
	content := ""

	params := make([]string, 0)
	for key, value := range *p.Params {
		switch v := value.(type) {
		case string:
			params = append(params, fmt.Sprintf("%s:%s", key, v))
		case int:
			params = append(params, fmt.Sprintf("%s:%d", key, v))
		case float64:
			params = append(params, fmt.Sprintf("%s:%f", key, v))
		case bool:
			params = append(params, fmt.Sprintf("%s:%t", key, v))
		default:
			params = append(params, fmt.Sprintf("%s:%s", key, v))
		}
	}

	content = fmt.Sprintf("%d;%s;", p.Timestamp.Unix(), strings.Join(params, ","))
	crc := calculateCrc([]byte(content))
	content = fmt.Sprintf("<Ps>%s%04X</Ps>", content, crc)
	return &content
}
